File: alpaca/templates/python/lib/http_client/auth_handler.py
# AuthHandler takes care of devising the auth type and using it
class AuthHandler():

	HTTP_PASSWORD = 0
	HTTP_TOKEN = 1
{{if .Api.authorization.oauth}}
	URL_SECRET = 2
	URL_TOKEN = 3
{{end}}
	def __init__(self, auth):
		self.auth = auth

	# Calculating the Authentication Type
	def get_auth_type(self):
		if 'username' in self.auth and 'password' in self.auth:
			return self.HTTP_PASSWORD
		elif 'http_token' in self.auth:
			return self.HTTP_TOKEN{{if .Api.authorization.oauth}}
		elif 'client_id' in self.auth and 'client_secret' in self.auth:
			return self.URL_SECRET
		elif 'access_token' in self.auth:
			return self.URL_TOKEN{{end}}
		else:
			return -1

	def set(self, request):
		if len(self.auth.keys()) == 0:
			return request

		auth = self.get_auth_type()

		if auth == HTTP_PASSWORD:
			request = http_password(request)
		elif auth == HTTP_TOKEN:
			request = http_token(request){{if .Api.authorization.oauth}}
		elif auth == URL_SECRET:
			request = url_secret(request)
		elif auth == URL_TOKEN:
			request = url_token(request){{end}}
		else:
			raise StandardError("Unable to calculate authorization method. Please check")

		return request

	# Basic Authorization with username and password
	def http_password(self, request):
		request['auth'] = (self.auth['username'], self.auth['password'])
		return request

	# Authorization with HTTP token
	def http_token(self, request):
		request['headers']['Authorization'] = 'token ' + self.auth['http_token']
		return request
{{if .Api.authorization.oauth}}
	# OAUTH2 Authorization with client secret
	def url_secret(self, request):
		request['params']['client_id'] = self.auth['client_id']
		request['params']['client_secret'] = self.auth['client_secret']
		return request

	# OAUTH2 Authorization with access token
	def url_token(self, request):
		request['params']['access_token'] = self.auth['access_token']
		return request
{{end}}

File: alpaca/templates/ruby/lib/http_client/auth_handler.rb
require "base64"

module {{.Pkg.name}}

  module HttpClient

    # AuthHandler takes care of devising the auth type and using it
    class AuthHandler < Faraday::Middleware

      HTTP_PASSWORD = 0
      HTTP_TOKEN = 1
{{if .Api.authorization.oauth}}
      URL_SECRET = 2
      URL_TOKEN = 3
{{end}}
      def initialize(app, auth = {}, options = {})
        @auth = auth
        super(app)
      end

      def call(env)
        if !@auth.empty?
          case get_auth_type

          when HTTP_PASSWORD
            env = http_password env
          when HTTP_TOKEN
            env = http_token env{{if .Api.authorization.oauth}}
          when URL_SECRET
            env = url_secret env
          when URL_TOKEN
            env = url_token env{{end}}
          else
            raise StandardError.new "Unable to calculate authorization method. Please check"
          end
        end

        @app.call(env)
      end

      # Calculating the Authentication Type
      def get_auth_type()
        if @auth.has_key?(:username) and @auth.has_key?(:password)
          return HTTP_PASSWORD
        elsif @auth.has_key?(:http_token)
          return HTTP_TOKEN{{if .Api.authorization.oauth}}
        elsif @auth.has_key?(:client_id) and @auth.has_key?(:client_secret)
          return URL_SECRET
        elsif @auth.has_key?(:access_token)
          return URL_TOKEN{{end}}
        else
          return -1
        end
      end

      # Basic Authorization with username and password
      def http_password(env)
        code = Base64.encode64 "#{@auth[:username]}:#{@auth[:password]}"

        env[:headers]["Authorization"] = "Basic #{code}"

        return env
      end

      # Authorization with HTTP token
      def http_token(env)
        env[:headers]["Authorization"] = "token #{@auth[:http_token]}"

        return env
      end
{{if .Api.authorization.oauth}}
      # OAUTH2 Authorization with client secret
      def url_secret(env)
        query = {
          client_id: @auth[:client_id],
          client_secret: @auth[:client_secret]
        }

        merge_query env, query
      end

      # OAUTH2 Authorization with access token
      def url_token(env)
        query = { access_token: @auth[:access_token] }

        merge_query env, query
      end
{{end}}
      def query_params(url)
        if url.query.nil? or url.query.empty?
          {}
        else
          Faraday::Utils.parse_query url.query
        end
      end

      def merge_query(env, query)
        query = query.update query_params(env[:url])

        env[:url].query = Faraday::Utils.build_query query

        return env
      end
    end

  end

end
